drivers/gpio/gpio_mcp23xxx.c

```diff
@@ -0,0 +1,489 @@
+	static struct mcp23xxx_drv_data mcp23##lowert##xx_##inst##_drvdata = {                     \
+		/* Default for registers according to datasheet */                                 \
+		.reg_cache.iodir = 0xFFFF, .reg_cache.ipol = 0x0,   .reg_cache.gpinten = 0x0,      \
+		.reg_cache.defval = 0x0,   .reg_cache.intcon = 0x0, .reg_cache.iocon = 0x0,        \
+		.reg_cache.gppu = 0x0,	   .reg_cache.intf = 0x0,   .reg_cache.intcap = 0x0,       \
+		.reg_cache.gpio = 0x0,	   .reg_cache.olat = 0x0,                                  \
+	};
+
+#define INST_DT_MCP23XXX(inst, t) DT_INST(inst, microchip_mcp23##t##xx)
+
+#define MCP23SXX_BUS(inst, t)                                                                      \
+	{                                                                                          \
+		.spi = SPI_DT_SPEC_GET(                                                            \
+			INST_DT_MCP23XXX(inst, t),                                                 \
+			SPI_OP_MODE_MASTER | SPI_MODE_CPOL | SPI_MODE_CPHA | SPI_WORD_SET(8), 0)   \
+	}
+
+#define MCP230XX_BUS(inst, t)                                                                      \
+	{                                                                                          \
+		.i2c = I2C_DT_SPEC_GET(INST_DT_MCP23XXX(inst, t))                                  \
+	}
+
+#define GPIO_MCP23XXX_DEVICE(inst, t, lowert)                                                      \
+	static struct mcp23xxx_config mcp23##lowert##xx_##inst##_config = {     \
+		.config = {					       \
+			.port_pin_mask =			       \
+				GPIO_PORT_PIN_MASK_FROM_DT_NODE(INST_DT_MCP23XXX(inst, lowert)), \
+		},						       \
+		.bus = MCP23##t##XX_BUS(inst, lowert), \
+		.ngpios =  DT_PROP(INST_DT_MCP23XXX(inst, lowert), ngpios),		       \
+		.read_fn = mcp23##lowert##xx_read_port_regs,              \
+		.write_fn = mcp23##lowert##xx_write_port_regs,              \
+		.bus_fn = mcp23##lowert##xx_bus_is_ready              \
+	};                  \
+	MCP23XXX_DRV_DATA(inst, lowert)                                                            \
+	DEVICE_DT_DEFINE(INST_DT_MCP23XXX(inst, lowert), mcp23xxx_init, NULL,                      \
+			 &mcp23##lowert##xx_##inst##_drvdata, &mcp23##lowert##xx_##inst##_config,  \
+			 POST_KERNEL, CONFIG_GPIO_MCP23##t##XX_INIT_PRIORITY, &api_table);
+
+#define GPIO_MCP230XX_DEVICE(n) GPIO_MCP23XXX_DEVICE(n, 0, 0)
+#define GPIO_MCP23SXX_DEVICE(n) GPIO_MCP23XXX_DEVICE(n, S, s)
+
+#define CALL_WITH_ARG(arg, expr) expr(arg);
+
+#define INST_DT_MCP23XXX_FOREACH(t, inst_expr)                                                     \
+	UTIL_LISTIFY(DT_NUM_INST_STATUS_OKAY(microchip_mcp23##t##xx), CALL_WITH_ARG, inst_expr)
+
+#ifdef CONFIG_GPIO_MCP230XX
+INST_DT_MCP23XXX_FOREACH(0, GPIO_MCP230XX_DEVICE);
+#endif
+
+#ifdef CONFIG_GPIO_MCP23SXX
+INST_DT_MCP23XXX_FOREACH(s, GPIO_MCP23SXX_DEVICE);
+#endif
```

Comment:
I think that this can be simplified if there's a source for each variant (i.e. each one takes care of its own init), and a common file with shared functionality. All this macro magic is hard to read (sometimes is unavoidable, so maybe I'm missing something)

Reply:
We could certainly just have a shared file and then dedicated driver files that leverage it and do their own init. I modeled this off the existing driver in the tree, that took this approach.

Reply:
@gmarull  Would you definitely like this refactored this way?

Reply:
I definitely think readability/quality of the driver would improve if all this macro magic can be avoided

Reply:
Ok, I'll take a stab at moving the shared functions out to a common file, and then separate driver files with the init logic for the mcp23sxx and mcp230xx.

Reply:
Updated to take this approach.